p2panda-net/src/sync/mod.rs

```diff
@@ -9,41 +9,52 @@ use std::sync::Arc;
 
 use anyhow::Result;
 use futures_util::{AsyncRead, AsyncWrite, SinkExt};
-use iroh_gossip::proto::TopicId;
 use iroh_net::key::PublicKey;
 use p2panda_sync::{FromSync, SyncError, SyncProtocol};
 use tokio::sync::mpsc;
 use tokio_util::sync::PollSender;
 use tracing::{debug, error};
 
-use crate::engine::ToEngineActor;
+use crate::{engine::ToEngineActor, Topic};
 
 /// Initiate a sync protocol session over the provided bi-directional stream for the given peer and
 /// topic.
-pub async fn initiate_sync<S: AsyncWrite + Send + Unpin, R: AsyncRead + Send + Unpin>(
+pub async fn initiate_sync<T, S, R>(
     mut send: &mut S,
     mut recv: &mut R,
     peer: PublicKey,
-    topic: TopicId,
-    sync_protocol: Arc<dyn for<'a> SyncProtocol<'a> + 'static>,
-    engine_actor_tx: mpsc::Sender<ToEngineActor>,
-) -> Result<()> {
+    topic: T,
+    sync_protocol: Arc<dyn for<'a> SyncProtocol<'a, T> + 'static>,
+    engine_actor_tx: mpsc::Sender<ToEngineActor<T>>,
+) -> Result<()>
+where
+    T: Topic + 'static,
+    S: AsyncWrite + Send + Unpin,
+    R: AsyncRead + Send + Unpin,
+{
     debug!(
         "initiate sync session with peer {} over topic {:?}",
         peer, topic
     );
 
     // Set up a channel for receiving new application messages.
-    let (tx, mut rx) = mpsc::channel(128);
+    let (tx, mut rx) = mpsc::channel::<FromSync<T>>(128);
     let mut sink = PollSender::new(tx).sink_map_err(|e| SyncError::Critical(e.to_string()));
 
     // Spawn a task which picks up any new application messages and sends them on to the engine
     // for handling.
     let mut sync_handshake_success = false;
+    let topic_id = topic.id();
     tokio::spawn(async move {
         while let Some(message) = rx.recv().await {
             // We expect the first message to be a topic id
-            if let FromSync::Topic(id) = &message {
+            // @TODO: We don't need to receive the topic from within the sync session as we
```

Comment:
Seems worthwhile to keep this in place for now but I'll defer to @adzialocha .

Reply:
Oh, haven't realized that this was also in the "initiator" logic, it shouldn't need to be there indeed as far as I can see it!

Reply:
> It is worth considering that the sync scope portion of a topic may have change since the network subscription occurred, if that is the correct behavior then this is actually a good way to receive the most current topic state.

I'm not sure I understand this / haven't considered this. How and where can the topic change? Do you mean between calling "subscribe" and the actual sync session happening?

Reply:
I was thinking about the fact that a topic `T` is stored in multiple locations, both internally inside network, and externally by the user. The internal version is passed into `subscribe` and from that point is never changed, the user stored copy is exposed to any sync protocol via a `TopicMap` and this _may_ be mutated without the networks knowledge. The point I'm making is that it's possible for a topic to get "out of sync" with itself... in a way... maybe this is more of an API documentation point: "if you change a topic in any way, even if the id remains the same, it's advised that `subscribe()` be re-called".

Reply:
> I was thinking about the fact that a topic `T` is stored in multiple locations, both internally inside network, and externally by the user. The internal version is passed into `subscribe` and from that point is never changed, the user stored copy is exposed to any sync protocol via a `TopicMap` and this _may_ be mutated without the networks knowledge. The point I'm making is that it's possible for a topic to get "out of sync" with itself... in a way... maybe this is more of an API documentation point: "if you change a topic in any way, even if the id remains the same, it's advised that `subscribe()` be re-called".

Interesting, I'd consider Topic almost like a messaging API a user defines themselves for telling other peers what they would like to have, and similar to an API like that the message doesn't change once its sent (or you have to re-send) if sth. changed.

I can't see in the code how that change might happen, can you walk me through an example in the code? Maybe there's a way to prevent it.

Reply:
It isn't exactly visible in our code, as it's more about use of the API. The main point is that calls to `subscribe` on network need to be coordinated with updates to a `TopicMap` (the newly subscribed topic needs to be present as a key here).

It somehow relates to our discussion around "what happens if I subscribe to two topics which have the same id" as defining the expected behaviour here is relevent.

Reply:
Thinking about this again, in many cases `TopicMap` will wrap calls to some store, and it will be able to interpret whatever topic it gets, and dynamically understand the scope. So this "out-of-sync" behaviour is actually a side-effect of only some (probably naive and niche) implementations.

Reply:
Turns out we actually want to wait in `initiate_sync`, but the message variant naming was confusing. I changed the name from `FromSync::Topic` to `FromSync::HandshakeSuccess` to better explain what we are waiting for. The contained topic is required in `accept_sync` but not here in `initiate_sync`. However in both cases we want to wait for the handshake to complete.